base/src/functions/logical.rs

```diff
@@ -66,91 +66,62 @@ impl Model {
-                                CalcResult::String(_value) => {
-                                    true_count += 1;
-                                }
-                                error @ CalcResult::Error { .. } => return error,
-                                CalcResult::Range { .. } => {}
-                                CalcResult::EmptyCell | CalcResult::EmptyArg => {}
-                            }
-                        }
-                    }
-                }
-                error @ CalcResult::Error { .. } => return error,
-                CalcResult::EmptyCell | CalcResult::EmptyArg => {}
-            };
-        }
-        if true_count == 0 {
-            return CalcResult::new_error(
-                Error::VALUE,
-                cell,
-                "Boolean values not found".to_string(),
-            );
+        fn fold_fn(acc: Option<bool>, value: bool) -> bool {
+            acc.unwrap_or(true) && value
         }
-        CalcResult::Boolean(true)
+        self.fn_logical_nary(args, cell, fold_fn, Some(false))
     }
 
     pub(crate) fn fn_or(&mut self, args: &[Node], cell: CellReferenceIndex) -> CalcResult {
+        fn fold_fn(acc: Option<bool>, value: bool) -> bool {
+            let r = acc.unwrap_or(false) || value;
+            println!("acc: {:?}, value: {:?}, r: {:?}", acc, value, r);
+            r
+        }
+        self.fn_logical_nary(args, cell, fold_fn, Some(true))
+    }
+
+    pub(crate) fn fn_xor(&mut self, args: &[Node], cell: CellReferenceIndex) -> CalcResult {
+        fn fold_fn(acc: Option<bool>, value: bool) -> bool {
+            acc.unwrap_or(false) ^ value
+        }
+        self.fn_logical_nary(args, cell, fold_fn, None)
+    }
+
+    /// Base function for AND, OR, XOR. These are all n-ary functions that perform a boolean operation on a series of
+    /// boolean values. These boolean values are sourced from `args`. Note that there is not a 1-1 relationship between
+    /// arguments and boolean values evaluated (see how Ranges are handled for example).
+    ///
+    /// Each argument in `args` is evaluated and the resulting value is interpreted as a boolean as follows:
+    /// - Boolean: The value is used directly.
+    /// - Number: 0 is FALSE, all other values are TRUE.
+    /// - Range: Each cell in the range is evaluated as if they were individual arguments with some caveats
+    /// - Empty arg: FALSE
+    /// - Empty cell & String: Ignored, behaves exactly like the argument wasn't passed in at all
+    /// - Error: Propagated
+    ///
+    /// If no arguments are provided, or all arguments are ignored, the function returns a #VALUE! error
+    ///
+    /// **`fold_fn`:** The function that combines the running result with the next value boolean value. The running result
+    /// starts as `None`.
+    ///
+    /// **`short_circuit_value`:** If the running result reaches `short_circuit_value`, the function returns early.
+    fn fn_logical_nary(
```

Comment:
Could you call this method `logical_nary` instead? Function names like `fn_or` means there is an Excel function called `OR`. Nothing relies on this but I'd like to keep that semantics if possible.